### PR TITLE
Fix webform validation conditions and refactor logic application contexts

### PR DESCRIPTION
**Pull request by tangrammer**
#### Before the PR (what is the issue or what needed to be done)

- Remove Cascade question type as not allowed
- Validate survey (and update webform boolean column db) only on `surveys/{id}/webform_id` and [publishing survey logic](https://github.com/akvo/akvo-flow/blob/issue-3562/GAE/src/org/waterforpeople/mapping/app/web/SurveyAssemblyServlet.java#L197) (thus webform python backend only uses s3 survey published definitions)

#### Reviewer Checklist
* [ ] Added an explanation about the work done
* [ ] Connected the PR and the issue on Zenhub
* [ ] Added a test plan to the issue
* [ ] Updated the copyright header (when relevant)
* [ ] Formatted the code
* [ ] Added a documentation (if relevant)
* [ ] Added some unit tests (if relevant)

### CODE REVIEW

File: GAE/src/com/gallatinsystems/survey/domain/WebForm.java [hunk 1]
```diff
@@ -25,17 +25,26 @@
...
         unsupportedTypes.add(Question.Type.SIGNATURE.toString());
         unsupportedTypes.add(Question.Type.CADDISFLY.toString());
         return unsupportedTypes;
     }
 
-    public static boolean validWebForm(final List<Question> questions){
-        
-        List<Question> validQuestions = questions.stream().filter(i -> !unsupportedQuestionTypes().contains(i.getType().toString())).collect(Collectors.toList());
+    public static boolean validQuestionGroups(final Survey survey){
+        return survey.getQuestionGroupMap().values().stream().filter(i -> i.getRepeatable().booleanValue()).collect(Collectors.toList()).size() == 0;
```
Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/domain/WebForm.java:35]: Why do we need an explicit `boolean` primitive? .... using `i.getRepeatable()` does the trick

File: GAE/src/com/gallatinsystems/survey/domain/WebForm.java [hunk 2]
```diff
@@ -25,17 +25,26 @@
...
+    }
 
+    public static boolean validSurveyGroup(final SurveyGroup surveyGroup){
+        return !surveyGroup.getMonitoringGroup().booleanValue();
```
Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/domain/WebForm.java:39]: Idem, that extra `.booleanValue()` is not required, I think

File: GAE/src/com/gallatinsystems/survey/domain/WebForm.java [hunk 3]
```diff
@@ -25,17 +25,26 @@
...
     }
 
-    public static boolean validWebForm(final List<Question> questions){
-        
-        List<Question> validQuestions = questions.stream().filter(i -> !unsupportedQuestionTypes().contains(i.getType().toString())).collect(Collectors.toList());
+    public static boolean validQuestionGroups(final Survey survey){
+        return survey.getQuestionGroupMap().values().stream().filter(i -> i.getRepeatable()).collect(Collectors.toList()).size() == 0;
+    }
 
+    public static Boolean validSurveyGroup(final SurveyGroup surveyGroup){
```
Comment (muloem) [inline GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]: So this one is a bit more subtle. It should be that webforms doesn't support monitoring forms. I.e. in a survey you have registration form and monitoring forms.  We should allow a registration form to be filled in.  Registration form id is === `SurveyGroup.newLocaleSurveyId` @marvinkome I guess this is already done on UI side?

Comment (marvinkome) [reply GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]: Yeah on the UI, monitoring surveys aren't supported

Comment (muloem) [reply GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]: Monitoring surveys or monitoring _forms_?

Comment (marvinkome) [reply GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]: 👀 monitoring survey groups

Comment (muloem) [reply GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]: Should be the same... in a monitoring Survey(Group) you have two kinds of forms, registration and monitoring forms. Registration forms can be shared via webforms, monitoring forms cannot be shared. Registration form id is found by looking at the `SurveyGroup.newLocaleSurveyId` property

File: GAE/src/com/gallatinsystems/survey/domain/WebForm.java [hunk 4]
```diff
@@ -25,17 +25,26 @@
...
+        return !surveyGroup.getMonitoringGroup();
+    }
+
+    public static boolean validWebForm(final SurveyGroup surveyGroup, final Survey survey, final List<Question> questions){
+        boolean validQuestionGroups = validQuestionGroups(survey);
+        if (!validQuestionGroups) return false;
```
Comment (iperdomo) [inline GAE/src/com/gallatinsystems/survey/domain/WebForm.java:44]: Let's always use `{ }` in `if` conditions.

File: GAE/src/com/gallatinsystems/survey/domain/WebForm.java [hunk 5]
```diff
@@ -23,19 +23,32 @@
...
     }
 
-    public static boolean validWebForm(final List<Question> questions){
-        
-        List<Question> validQuestions = questions.stream().filter(i -> !unsupportedQuestionTypes().contains(i.getType().toString())).collect(Collectors.toList());
+    public static boolean validQuestionGroups(final Survey survey) {
+        return survey.getQuestionGroupMap().values().stream().filter(i -> i.getRepeatable()).collect(Collectors.toList()).size() == 0;
+    }
 
+    public static boolean validSurveyGroup(final Survey survey, final SurveyGroup surveyGroup) {
```
Comment (muloem) [inline GAE/src/com/gallatinsystems/survey/domain/WebForm.java:38]:
```suggestion
    public static boolean validForm(final Survey survey, final SurveyGroup surveyGroup) {
```

**Pull request merged**